## Notebook to test out profiling

To profile my backtracking algorithm I modified the `main()` in my `main.py` file so that it doesn't take any command line arguments.

Here is the modified function:

```python
import cProfile

def profiled_main(sudoku_path):
    """
    Profiled version of main function that handles the execution of the Sudoku solver program.
    Note: This function is used for profiling the program using cProfile and therefore does not handle user input.

    Parameters
    ----------
        sudoku_path (str): Path to the input sudoku file.

    Returns
    ----------
        str: The solved sudoku.
    """
    with open(sudoku_path, "r") as f:
        input_sudoku = f.read()

    sudoku_board = parse_grid(input_sudoku)
    print(f"Uploaded sudoku:\n\n{input_sudoku}")
    solved_sudoku = solveBacktrack(sudoku_board, 0, 0)
    print(f"Solved sudoku:\n\n{displaySudoku(solved_sudoku)}")
    return 0
```

The function is the called using the following:

```python
if __name__ == "__main__":
    input_file_path = "test/example_sudokus/hard_sudoku2.txt"
    cProfile.run("profiled_main(input_file_path)")
```

Results using `hard_sudoku1.txt`:

```txt
380|000|000
000|400|785
009|020|300
---+---+---
060|090|000
800|302|009
000|040|070
---+---+---
001|070|500
495|006|000
000|000|092
```

```bash
         63146 function calls (57406 primitive calls) in 0.169 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <frozen codecs>:260(__init__)
        1    0.000    0.000    0.169    0.169 <string>:1(<module>)
   5741/1    0.022    0.000    0.163    0.163 backtracking.py:117(solveBacktrack)
    51425    0.118    0.000    0.118    0.000 backtracking.py:14(validateCell)
     5741    0.022    0.000    0.022    0.000 backtracking.py:95(findEmptyCell)
        1    0.000    0.000    0.000    0.000 cp1252.py:22(decode)
        1    0.000    0.000    0.169    0.169 main.py:161(profiled_main)
        1    0.000    0.000    0.000    0.000 utils.py:155(displaySudoku)
        1    0.000    0.000    0.000    0.000 utils.py:9(parse_grid)
        1    0.000    0.000    0.000    0.000 {built-in method _codecs.charmap_decode}
        1    0.000    0.000    0.000    0.000 {built-in method _io.open}
        1    0.000    0.000    0.169    0.169 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
       11    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        2    0.006    0.003    0.006    0.003 {built-in method builtins.print}
        1    0.000    0.000    0.000    0.000 {method '__exit__' of '_io._IOBase' objects}
      213    0.000    0.000    0.000    0.000 {method 'append' of 'list' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 {method 'read' of '_io.TextIOWrapper' objects}
```

Results using `hard_sudoku2.txt`:

```text
850|002|400
720|000|009
004|000|000
---+---+---
000|107|002
305|000|900
040|000|000
---+---+---
000|080|070
017|000|000
000|036|040
```

```bash
         3692014 function calls (3356377 primitive calls) in 9.045 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.000    0.000 <frozen codecs>:260(__init__)
        1    0.000    0.000    9.045    9.045 <string>:1(<module>)
 335638/1    1.194    0.000    9.042    9.042 backtracking.py:117(solveBacktrack)
  3020499    6.524    0.000    6.524    0.000 backtracking.py:14(validateCell)
   335638    1.324    0.000    1.324    0.000 backtracking.py:95(findEmptyCell)
        1    0.000    0.000    0.000    0.000 cp1252.py:22(decode)
        1    0.000    0.000    9.045    9.045 main.py:161(profiled_main)
        1    0.000    0.000    0.000    0.000 utils.py:155(displaySudoku)
        1    0.000    0.000    0.000    0.000 utils.py:9(parse_grid)
        1    0.000    0.000    0.000    0.000 {built-in method _codecs.charmap_decode}
        1    0.000    0.000    0.000    0.000 {built-in method _io.open}
        1    0.000    0.000    9.045    9.045 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}
       11    0.000    0.000    0.000    0.000 {built-in method builtins.len}
        2    0.002    0.001    0.002    0.001 {built-in method builtins.print}
        1    0.000    0.000    0.000    0.000 {method '__exit__' of '_io._IOBase' objects}
      213    0.000    0.000    0.000    0.000 {method 'append' of 'list' objects}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
        1    0.000    0.000    0.000    0.000 {method 'read' of '_io.TextIOWrapper' objects}
```

From the above, the backtracking algorithm is what is slowing the speed of my program. I now investigate the backtracking algorithm itself

In [93]:
from backtracking import solveBacktrack, validateCell, findEmptyCell
from utils import parse_grid

with open("../test/example_sudokus/hard_sudoku1.txt", "r") as f:
    hard_sudoku1 = f.read()

with open("../test/example_sudokus/hard_sudoku2.txt", "r") as f:
    hard_sudoku2 = f.read()

with open("../test/example_sudokus/hard_sudoku3.txt", "r") as f:
    hard_sudoku3 = f.read()

hard_sudoku1 = parse_grid(hard_sudoku1)
hard_sudoku2 = parse_grid(hard_sudoku2)
hard_sudoku3 = parse_grid(hard_sudoku3)

Profiling `backtracking.py` gives:

```bash
(c1_coursework_wp289) C:\Users\William Purvis\OneDrive - University of Cambridge\DIS\Michealmas\RC_coursework\wp289>python src/backtracking.py
         3691778 function calls (3356141 primitive calls) in 8.904 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    8.904    8.904 <string>:1(<module>)
 335638/1    1.183    0.000    8.904    8.904 backtracking.py:118(solveBacktrack)
  3020499    6.408    0.000    6.408    0.000 backtracking.py:15(validateCell)
   335638    1.313    0.000    1.313    0.000 backtracking.py:96(findEmptyCell)
        1    0.000    0.000    8.904    8.904 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

I then downloaded 1 million sudoku boards from [kaggle](https://www.kaggle.com/datasets/bryanpark/sudoku/) and randomly sampled 1000 boards to test my algorithm on.

Here is the code I used to get 1000 boards in the correct format:

```python
import pandas as pd
import numpy as np
import pickle

np.random.seed(42)

all_sudoku = pd.read_csv('sudoku.csv', header=None)
quizzes = all_sudoku.iloc[1:, 0]

# randomly select 1000 sudoku
test_quizzes = quizzes.sample(n=1000)
test_quizzes = np.array(test_quizzes)

sudoku_boards = [np.array(list(map(int, quiz))).reshape(9, 9).tolist() 
                for quiz in test_quizzes]

with open('test_sudoku_boards.pkl', 'wb') as f:
    pickle.dump(sudoku_boards, f)
```

### Profiling current backtracking implementation (V1.0.0)

In this next section, the 1.0.0 version of my algorithm is tested on the 10000 sudokus from kaggle

In [36]:
import pickle
import time
import copy
from backtracking import solveBacktrack

# load 1000 sudokus
with open("../test/example_sudokus/tenk_sudoku_boards.pkl", "rb") as f:
    loaded_boards = pickle.load(f)

# run solver on 1000 boards
time_to_solve = []
for board in loaded_boards:
    board_copy = copy.deepcopy(board)
    start_time = time.time()
    sudoku1_solved = solveBacktrack(board_copy, 0, 0)
    end_time = time.time()
    time_to_solve.append(end_time - start_time)

In [40]:
import numpy as np

print(f"Average time to solve 10,000 boards: {np.mean(time_to_solve):.6f}")
print(f"Standard deviation: {np.std(time_to_solve):.6f}")

Average time to solve 10,000 boards: 0.001658
Standard deviation: 0.004571


In [55]:
hard_sudoku2_copy = copy.deepcopy(hard_sudoku2)
hard_sudoku3_copy = copy.deepcopy(hard_sudoku3)

start_time = time.time()
sudoku3_solved = solveBacktrack(hard_sudoku3_copy, 81, 81)
end_time = time.time()
print(f"Time to solve hard sudoku 3: {end_time - start_time:.6f}")

Time to solve hard sudoku 3: 2320.798187


In [57]:
start_time = time.time()
sudoku2_solved = solveBacktrack(hard_sudoku2_copy, 81, 81)
end_time = time.time()
print(f"Time to solve hard sudoku 2: {end_time - start_time:.6f} s")

Time to solve hard sudoku 2: 4.518573 s


Using MRV heuristics + backtracking on `hard_sudoku2`:

```txt
Solved sudoku in 1.8709 seconds. using MRV
Solved sudoku in 4.0468 seconds. without MRV
```

`hard_sudoku3` took `2320.798187` seconds to solve without MRV. Using MRV this reduces to:

```txt
Solved sudoku in 20.3896 seconds. using MRV
```

Using `cProfile` on `hard_sudoku2.txt`.

Without MRV:

```txt
         3691778 function calls (3356141 primitive calls) in 11.029 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   11.029   11.029 <string>:1(<module>)
   335638    1.712    0.000    1.712    0.000 backtracking.py:100(findEmptyCell)
 335638/1    1.478    0.000   11.029   11.029 backtracking.py:124(solveBacktrack)
  3020499    7.838    0.000    7.838    0.000 backtracking.py:18(validateCell)
        1    0.000    0.000   11.029   11.029 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

With MRV:

```txt
         1419837 function calls (1415801 primitive calls) in 4.559 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    4.559    4.559 <string>:1(<module>)
     4037    0.092    0.000    4.445    0.001 backtracking2.py:117(find_empty_cell_MRV)
   4037/1    0.021    0.000    4.559    4.559 backtracking2.py:153(solve_backtrack_MRV)
  1274193    3.960    0.000    3.960    0.000 backtracking2.py:8(validate_cell)
   137567    0.486    0.000    4.353    0.000 backtracking2.py:90(n_possible_values)
        1    0.000    0.000    4.559    4.559 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

Using `cProfile` on `hard_sudoku3.txt`.

Without MRV:

```txt
        760928234 function calls (691752918 primitive calls) in 2320.798 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000 2132.140 2132.140 <string>:1(<module>)
 69175317  320.649    0.000  320.649    0.000 backtracking.py:100(findEmptyCell)
69175317/1  283.327    0.000 2132.140 2132.140 backtracking.py:124(solveBacktrack)
622577597 1528.164    0.000 1528.164    0.000 backtracking.py:18(validateCell)
        1    0.000    0.000 2132.140 2132.140 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

With MRV

```txt
         16346635 function calls (16301368 primitive calls) in 42.844 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000   42.843   42.843 <string>:1(<module>)
    45268    0.781    0.000   41.799    0.001 backtracking2.py:117(find_empty_cell_MRV)
  45268/1    0.173    0.000   42.843   42.843 backtracking2.py:153(solve_backtrack_MRV)
 14671202   37.379    0.000   37.379    0.000 backtracking2.py:8(validate_cell)
  1584894    4.511    0.000   41.019    0.000 backtracking2.py:90(n_possible_values)
        1    0.000    0.000   42.844   42.844 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

Running backtracking and backtracking + MRV on 10,000 sudokus:

```txt
(c1_coursework_wp289) C:\Users\William Purvis\OneDrive - University of Cambridge\DIS\Michealmas\RC_coursework\wp289>python src/profiling.py
100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10000/10000 [00:16<00:00, 596.71it/s]
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10000/10000 [00:00<00:00, 138893.90it/s] 
Average time to solve sudoku with backtracking:
0.001666 seconds
Standard deviation: 0.002081 seconds
Average time to solve sudoku with backtracking and MRV heuristics:
0.000007 seconds
Standard deviation: 0.000083 seconds

```

Using cython on `hard_sudoku2` and `hard_sudoku3`:

```txt
Time taken to solve hard_sudoku2: 0.604491
Time taken to solve hard_sudoku3: 5.314255
```
Using cProfile for `hard_sudoku3`:

```txt
         3 function calls in 5.403 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    5.403    5.403    5.403    5.403 <string>:1(<module>)
        1    0.000    0.000    5.403    5.403 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
```

I then changed the input for my cython functions from list of lists which are native to python, to c arrays. This improved performance even further:

```txt
Time taken to solve hard_sudoku2: 0.075638
Time taken to solve hard_sudoku3: 1.051183
```

Using cProfile for `hard_sudoku3`:

```txt
         3 function calls in 0.815 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.815    0.815    0.815    0.815 <string>:1(<module>)
        1    0.000    0.000    0.815    0.815 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}

```